lib/client-assets.php

```diff
@@ -565,3 +565,78 @@ function gutenberg_register_vendor_scripts( $scripts ) {
 // Enqueue stored styles.
 add_action( 'wp_enqueue_scripts', 'gutenberg_enqueue_stored_styles' );
 add_action( 'wp_footer', 'gutenberg_enqueue_stored_styles', 1 );
+
+
+/**
+ * Add footnotes to the end of the content.
+ *
+ * @param string $content Content.
+ *
+ * @return string Content with footnotes.
+ */
+function gutenberg_footnotes_the_content( $content ) {
+	if ( ! strpos( $content, '<data' ) ) {
+		return $content;
+	}
+
+	$notes = array();
+
+	// To be replaced later with a dynamic data API:
+	// replaceDataByType( 'core/footnote', ( attributes ) => <sup>...</sup> );
+	$content = preg_replace_callback(
+		'/<data value=\"core\/footnote\">\[(.*?)\]<\/data>/',
+		function( $matches ) use ( &$notes ) {
+			$note = $matches[1];
+			$id = md5( $note );
+
+			if ( isset( $notes[ $id ] ) ) {
+				$notes[ $id ]['count'] += 1;
+			} else {
+				$notes[ $id ] = array(
+					'note'  => $note,
+					'count' => 1,
+				);
+			}
+
+			// List starts at 1. If the note already exists, use the existing index.
+			$index = 1 + array_search( $id, array_keys( $notes ), true );
+			$count = $notes[ $id ]['count'];
+
+			return (
+				'<sup><a class="note-link" href="#' . $id . '" id="' . $id . '-link-' . $count . '">[' . $index . ']</a></sup>'
```

Comment:
Musing a bit here to explore ways this can be done with the HTML API. This is me playing with the idea and seeing how this need might be taken care of and lead the HTML API design…

This first one uses the Tag Processor in a way that's tightly coupled to internal details. It's not the way we want people to use the Tag Processor, but it's technically possible with what's in Core.

```php
class FootNoteProcessor extends WP_HTML_Tag_Processor {
	public function replace_footnotes() {
		$notes = array();

		while ( $this->next_tag( 'data' ) ) {
			if ( 'core/footnote' !== $this->get_attribute( 'value' ) ) {
				continue;
			}

			// This bookmark stuff is internal. Possible to use it here now
			// like this, but not something intended to be used in practice
			// outside of the Processor classes.
			$this->set_bookmark( 'start' );
			$start = $this->bookmarks['start']->start;

			// Move to the very next tag token. If it isn't closing
			// the DATA tag then bail; it's not what we expect.
			if ( ! (
				$this->next_tag( array( 'tag_closers' => 'visit' ) ) &&
				'DATA' === $this->get_tag() &&
				$this->is_tag_closer()
			) ) {
				return false;
			}
			$this->set_bookmark( 'end' );

			// Grab the existing inner content.
			$note = substr(
				$this->html,
				$this->bookmarks['start']->end,
				$this->bookmarks['end']->start
			);

			$id = md5( $note );

			if ( isset( $notes[ $id ] ) ) {
				$notes[ $id ]['count'] += 1;
			} else {
				$notes[ $id ] = array(
					'note'  => $note,
					'count' => 1,
				);
			}

			// List starts at 1. If the note already exists, use the existing index.
			$index = 1 + array_search( $id, array_keys( $notes ), true );
			$count = $notes[ $id ]['count'];


			$footnote = new WP_HTML_Tag_Processor( "<sup><a>[{$index}]</a></sup>" );
			$footnote->next_tag( 'a' );
			$footnote->add_class( 'note-link' );
			$footnote->set_attribute( 'href', "#{$id}" );
			$footnote->set_attribute( 'id', "{$id}-link-{$count}" );

			// Replace the outer content with the new structure.
			$this->lexical_updates[] = new WP_HTML_Text_Replacement(
				$this->bookmarks['start']->start,
				$this->bookmarks['end']->end + 1,
				$footnote->get_updated_html()
			);

			$this->bookmarks['start']->start = $start;
			$this->bookmarks['start']->end   = $start;
			$this->seek( 'start' );
			$this->next_tag();
		}

		if ( empty( $notes ) ) {
			return true;
		}

		…
	}
}

$footnote_processor = new FootnoteProcessor( $block_html );
$footnote_processor->replace_footnotes();
$with_footnotes = $footnote_processor->get_updated_html();
```

The HTML processor should end up making this far less noisy and remove the internal coupling, but isn't available yet. I think that the code above will work today.

```php
function replace_footnotes( $html ) {
	$p = new WP_HTML_Processor( $html );g
	$notes = array();

	while ( $p->select( 'data[value = "core/footnote"]' ) ) {
		$note = $p->get_inner_html();

		…

		$p->set_outer_html( $footnote->get_updated_html() );
	}

	if ( empty( $notes ) ) {
		return $p->get_updated_html();
	}

	…
}
```

And yesterday while exploring some unrelated bits I thought of an idea to use today's Tag Processor to generate the template in easier way, utilizing the funky HTML placeholder syntax I realized is possible through abuse of the "tag closer with invalid tag name is a comment" rule.

```php
$footnote = wp_html_template(
	<<<HTML
	<sup>
		<a class="note-link" href="#</%id>" id="</%id>-link-</%count>">[</%index>]</a>
	</sup>
HTML,
	array(
		'id'    => $id,
		'index' => $index,
		'count' => $count
	)
);
```